lib/stack_master/aws_driver/s3.rb

```diff
@@ -0,0 +1,73 @@
+require 'digest/md5'
+
+module StackMaster
+  module AwsDriver
+    class S3ConfigurationError < StandardError; end
+
+    class S3
+      def set_region(region)
+        @region = region
+        @s3 = nil
+      end
+
+      def upload_files(options)
+        bucket = options.fetch(:bucket) { raise StackMaster::AwsDriver::S3ConfigurationError, 'A bucket must be specified in order to use S3' }
+        prefix = options[:prefix]
+        bucket_region = options.fetch(:region, @region)
+        files = options.fetch(:files, [])
+
+        return if files.empty?
+
+        set_region(bucket_region) if bucket_region
+
+        current_objects = s3.list_objects(
+          prefix: prefix,
+          bucket: bucket
+        ).map(&:contents).flatten.inject({}){|h,obj|
+          h.merge(obj.key => obj)
+        }
+
+        files.each do |template, file|
+          body = file[:body]
+          file = file[:path]
+          key = template.dup
+          key.prepend("#{prefix}/") if prefix
+          file_md5 = Digest::MD5.file(file).to_s
+          body_md5 = Digest::MD5.hexdigest(body).to_s
+          s3_md5 = current_objects[key] ? current_objects[key].etag.gsub("\"", '') : nil
+
+          # require 'pry'
+          # binding.pry
+          unless file_md5 == s3_md5 or body_md5 == s3_md5
```

Comment:
Short version, it allows the bucket to contain pretty json and/or be synced up to a bucket separately from stack_master.

Long version, this gets a bit toward my own personal use case... at Logicworks, we have a library of CF templates that we use across a wide variety of projects. We have a CF bucket that we manage and update via CI/CD that is read only across a variety of different AWS users/accounts. So in our case, what we'll be doing is syncing that bucket down to the templates/ folder in stack_master.  As a result, the files in the bucket will not actually be writable most of the time, and we're relying on the MD5 of both places to be equal and prevent stack_master from even attempting to rewrite the contents of the bucket. 

And, since the content of body_md5 is the compressed json, and the contents or the bucket are a form of 'pretty' json, I'd prefer stack_master to accept either the pretty of the compressed version of the json to be acceptable.

how about:

````ruby
next if [body_md5, file_md5].include?(s3_md5)
````